src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataFilter.cs

```diff
@@ -29,21 +29,22 @@ public void OnResourceExecuting(ResourceExecutingContext context)
         /// <inheritdoc />
         public void OnResourceExecuted(ResourceExecutedContext context)
         {
-            _factory.GetTempData(context.HttpContext).Save();
         }
 
         /// <inheritdoc />
         public void OnResultExecuting(ResultExecutingContext context)
         {
+            if (context.Result is IKeepTempDataResult)
+            {
+                _factory.GetTempData(context.HttpContext).Keep();
+            }
+
+            _factory.GetTempData(context.HttpContext).Save();
```

Comment:
I see what you are saying. Looks like we missed some test coverage for this scenario. I now understand why you were mentioning about using `OnSendingHeaders` (or `OnStarting`). Using this would help us to resolve this issue. Will update the PR. Thanks!